File: sdks/csharp/PgChangeFeed.Client/Http/PgChangeFeedHttpClient.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using PgChangeFeed.Client.Http.Models;

namespace PgChangeFeed.Client.Http;

/// <summary>
/// Client for the PG Change Feed HTTP/JSON API: one method per capability —
/// <c>RegisterConsumerAsync</c>, <c>AcknowledgeConsumerAsync</c>,
/// <c>GetConsumerPositionAsync</c>, <c>RemoveConsumerAsync</c>,
/// <c>EnableTableAsync</c>, <c>DisableTableAsync</c>, <c>GetStatusAsync</c>,
/// <c>ListTablesAsync</c>, <c>RunRetentionAsync</c> and
/// <c>ReadChangesAsync</c>. Requests and responses are typed DTOs that mirror
/// the JSON documents of the API (<c>PgChangeFeed.Client.Http.Models</c>);
/// every non-success response becomes a typed <see cref="PgChangeFeedException"/>
/// subtype instead of a result type mixed with the success path. Connection
/// errors and timeouts of the transport are not converted; they reach the
/// caller as the <see cref="HttpRequestException"/> or
/// <see cref="TaskCanceledException"/> of the <see cref="HttpClient"/>.
///
/// The <see cref="HttpClient"/> is passed in, not owned — the caller controls
/// its lifetime, connection pooling and any <c>DelegatingHandler</c>
/// pipeline (proxies, retries, logging); this type never disposes it. The
/// bearer token and server address come from <see cref="PgChangeFeedClientOptions"/>,
/// supplied at construction — there is no global or static state, so a
/// process can hold several independently configured instances at once.
/// </summary>
public sealed class PgChangeFeedHttpClient
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    private readonly HttpClient _httpClient;
    private readonly PgChangeFeedClientOptions _options;

    /// <summary>Creates the client over an <see cref="HttpClient"/> the caller owns.</summary>
    /// <param name="httpClient">The client used for every request.</param>
    /// <param name="options">The server address and the bearer token.</param>
    public PgChangeFeedHttpClient(HttpClient httpClient, PgChangeFeedClientOptions options)
    {
        ArgumentNullException.ThrowIfNull(httpClient);
        ArgumentNullException.ThrowIfNull(options);
        _httpClient = httpClient;
        _options = options;
    }

    /// <summary>
    /// Registers a consumer, a named reader whose position the server keeps
    /// (<c>POST /consumers</c>, admin token). Registering an existing consumer
    /// changes nothing; the response reports it with <c>AlreadyRegistered</c>.
    /// </summary>
    public Task<RegisterConsumerResponse> RegisterConsumerAsync(
        RegisterConsumerRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return PostAsync<RegisterConsumerRequest, RegisterConsumerResponse>(
            "/consumers", request, cancellationToken);
    }

    /// <summary>
    /// Stores the position up to which a consumer has processed a source
    /// (<c>POST /consumers/acknowledge</c>, admin token). Repeating the stored
    /// position has no effect; an earlier position, or a position of another
    /// source, is rejected with <see cref="PgChangeFeedBadRequestException"/>.
    /// </summary>
    public Task<AcknowledgeConsumerResponse> AcknowledgeConsumerAsync(
        AcknowledgeConsumerRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return PostAsync<AcknowledgeConsumerRequest, AcknowledgeConsumerResponse>(
            "/consumers/acknowledge", request, cancellationToken);
    }

    /// <summary>
    /// Reads the stored position of a consumer (<c>GET /consumers/position</c>,
    /// reader or admin token). <c>Acknowledged</c> is <c>false</c> for a
    /// consumer that never acknowledged; <c>Offset</c> is then the starting
    /// position.
    /// </summary>
    public Task<ConsumerPositionResponse> GetConsumerPositionAsync(
        string consumerId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(consumerId);
        var query = BuildQuery(("consumer_id", consumerId));
        return GetAsync<ConsumerPositionResponse>($"/consumers/position?{query}", cancellationToken);
    }

    /// <summary>
    /// Removes a consumer (<c>POST /consumers/remove</c>, admin token);
    /// <c>Removed</c> is <c>false</c> for one that was never registered.
    /// </summary>
    public Task<RemoveConsumerResponse> RemoveConsumerAsync(
        string consumerId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(consumerId);
        return PostAsync<RemoveConsumerRequest, RemoveConsumerResponse>(
            "/consumers/remove", new RemoveConsumerRequest(consumerId), cancellationToken);
    }

    /// <summary>
    /// Starts capturing a table of a source (<c>POST /tables/enable</c>, admin
    /// token). <c>AlreadyEnabled</c> is <c>true</c> when the table was captured
    /// already; a table that does not exist in the source database raises
    /// <see cref="PgChangeFeedNotFoundException"/>.
    /// </summary>
    public Task<EnableTableResponse> EnableTableAsync(
        EnableTableRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return PostAsync<EnableTableRequest, EnableTableResponse>(
            "/tables/enable", request, cancellationToken);
    }

    /// <summary>
    /// Stops capturing a table (<c>POST /tables/disable</c>, admin token).
    /// <c>Retained</c> is <c>true</c> when changes already stored for the table
    /// remain readable; a table that does not exist in the source database
    /// raises <see cref="PgChangeFeedNotFoundException"/>.
    /// </summary>
    public Task<DisableTableResponse> DisableTableAsync(
        DisableTableRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return PostAsync<DisableTableRequest, DisableTableResponse>(
            "/tables/disable", request, cancellationToken);
    }

    /// <summary>
    /// Tells whether a table is captured (<c>Enabled</c>) or no longer captured
    /// with stored changes remaining (<c>Retained</c>)
    /// (<c>GET /tables/status</c>, reader or admin token). A table that was never
    /// enabled reports both as <c>false</c>; a table that does not exist in the
    /// source database raises <see cref="PgChangeFeedNotFoundException"/>.
    /// </summary>
    public Task<TableStatusResponse> GetStatusAsync(
        string source, string schema, string table, string publication, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(source);
        ArgumentException.ThrowIfNullOrWhiteSpace(schema);
        ArgumentException.ThrowIfNullOrWhiteSpace(table);
        ArgumentException.ThrowIfNullOrWhiteSpace(publication);
        var query = BuildQuery(
            ("source", source), ("schema", schema), ("table", table), ("publication", publication));
        return GetAsync<TableStatusResponse>($"/tables/status?{query}", cancellationToken);
    }

    /// <summary>
    /// Lists the captured tables and the tables that are no longer captured but
    /// whose stored changes remain (<c>GET /tables</c>, reader or admin token).
    /// </summary>
    public Task<ListTablesResponse> ListTablesAsync(
        string source, string publication, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(source);
        ArgumentException.ThrowIfNullOrWhiteSpace(publication);
        var query = BuildQuery(("source", source), ("publication", publication));
        return GetAsync<ListTablesResponse>($"/tables?{query}", cancellationToken);
    }

    /// <summary>
    /// Deletes the stored changes of a source that are older than
    /// <c>MinAgeNanos</c> and that every consumer with a stored position has
    /// passed (<c>POST /retention/run</c>, admin token); <c>Deleted</c> is the
    /// number removed.
    /// </summary>
    public Task<RunRetentionResponse> RunRetentionAsync(
        RunRetentionRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return PostAsync<RunRetentionRequest, RunRetentionResponse>(
            "/retention/run", request, cancellationToken);
    }

    /// <summary>
    /// Reads stored changes of a source (<c>GET /changes</c>, reader or admin
    /// token). <paramref name="source"/> is mandatory; <paramref name="schema"/>/
    /// <paramref name="table"/> are optional and independent;
    /// <paramref name="from"/> is inclusive, <paramref name="to"/> exclusive
    /// (both <c>commit_position</c> values). <paramref name="limit"/> cuts rows,
    /// not positions, and there is no default limit. A range without changes
    /// returns an empty list.
    /// </summary>
    public Task<ReadChangesResponse> ReadChangesAsync(
        string source,
        string? schema = null,
        string? table = null,
        long? from = null,
        long? to = null,
        int? limit = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(source);
        var query = BuildQuery(
            ("source", source),
            ("schema", schema),
            ("table", table),
            ("from", from?.ToString()),
            ("to", to?.ToString()),
            ("limit", limit?.ToString()));
        return GetAsync<ReadChangesResponse>($"/changes?{query}", cancellationToken);
    }

    private static string BuildQuery(params (string Key, string? Value)[] parameters)
    {
        var pairs = parameters
            .Where(p => p.Value is not null)
            .Select(p => $"{p.Key}={Uri.EscapeDataString(p.Value!)}");
        return string.Join('&', pairs);
    }

    private Task<TResponse> GetAsync<TResponse>(string pathAndQuery, CancellationToken cancellationToken)
        => SendAsync<TResponse>(HttpMethod.Get, pathAndQuery, content: null, cancellationToken);

    private Task<TResponse> PostAsync<TRequest, TResponse>(
        string path, TRequest request, CancellationToken cancellationToken)
    {
        var json = JsonSerializer.Serialize(request, JsonOptions);
        var content = new StringContent(json, Encoding.UTF8, "application/json");
        return SendAsync<TResponse>(HttpMethod.Post, path, content, cancellationToken);
    }

    private async Task<TResponse> SendAsync<TResponse>(
        HttpMethod method, string pathAndQuery, HttpContent? content, CancellationToken cancellationToken)
    {
        var uri = new Uri(_options.Address, pathAndQuery);
        using var request = new HttpRequestMessage(method, uri) { Content = content };
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _options.ApiToken);

        using var response = await _httpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
        var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);

        if (!response.IsSuccessStatusCode)
        {
            throw BuildException((int)response.StatusCode, body);
        }

        return DeserializeSuccessBody<TResponse>((int)response.StatusCode, body);
    }

    private static TResponse DeserializeSuccessBody<TResponse>(int statusCode, string body)
    {
        TResponse? result;
        try
        {
            result = JsonSerializer.Deserialize<TResponse>(body, JsonOptions);
        }
        catch (JsonException ex)
        {
            throw new PgChangeFeedMalformedResponseException(
                statusCode,
                $"PG Change Feed HTTP API returned status {statusCode} with a response body that is not " +
                "valid JSON.",
                ex);
        }

        return result
            ?? throw new PgChangeFeedMalformedResponseException(
                statusCode,
                $"PG Change Feed HTTP API returned status {statusCode} with an empty or " +
                "null response body.");
    }

    private static PgChangeFeedException BuildException(int statusCode, string body)
    {
        var message = ExtractErrorMessage(body);
        return statusCode switch
        {
            400 => new PgChangeFeedBadRequestException(statusCode, message),
            401 => new PgChangeFeedUnauthorizedException(statusCode, message),
            403 => new PgChangeFeedForbiddenException(statusCode, message),
            404 => new PgChangeFeedNotFoundException(statusCode, message),
            500 => new PgChangeFeedServerErrorException(statusCode, message),
            _ => new PgChangeFeedUnexpectedStatusException(statusCode, message),
        };
    }

    private static string ExtractErrorMessage(string body)
    {
        try
        {
            var error = JsonSerializer.Deserialize<ErrorResponse>(body, JsonOptions);
            return error?.Error ?? body;
        }
        catch (JsonException)
        {
            return body;
        }
    }
}

File: sdks/kotlin/pgchangefeed-kotlin/src/main/kotlin/io/github/pt9912/pgchangefeed/http/PgChangeFeedHttpClient.kt
package io.github.pt9912.pgchangefeed.http

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import io.github.pt9912.pgchangefeed.PgChangeFeedClientOptions
import io.github.pt9912.pgchangefeed.http.model.AcknowledgeConsumerRequest
import io.github.pt9912.pgchangefeed.http.model.AcknowledgeConsumerResponse
import io.github.pt9912.pgchangefeed.http.model.ConsumerPositionResponse
import io.github.pt9912.pgchangefeed.http.model.DisableTableRequest
import io.github.pt9912.pgchangefeed.http.model.DisableTableResponse
import io.github.pt9912.pgchangefeed.http.model.EnableTableRequest
import io.github.pt9912.pgchangefeed.http.model.EnableTableResponse
import io.github.pt9912.pgchangefeed.http.model.ErrorResponse
import io.github.pt9912.pgchangefeed.http.model.ListTablesResponse
import io.github.pt9912.pgchangefeed.http.model.ReadChangesResponse
import io.github.pt9912.pgchangefeed.http.model.RegisterConsumerRequest
import io.github.pt9912.pgchangefeed.http.model.RegisterConsumerResponse
import io.github.pt9912.pgchangefeed.http.model.RemoveConsumerResponse
import io.github.pt9912.pgchangefeed.http.model.RunRetentionRequest
import io.github.pt9912.pgchangefeed.http.model.RunRetentionResponse
import io.github.pt9912.pgchangefeed.http.model.TableStatusResponse
import java.net.http.HttpClient
import java.nio.charset.StandardCharsets

/**
 * Client for the PG Change Feed HTTP/JSON API: one method per capability —
 * `registerConsumer`, `acknowledgeConsumer`, `getConsumerPosition`,
 * `removeConsumer`, `enableTable`, `disableTable`, `getStatus`, `listTables`,
 * `runRetention` and `readChanges`. Requests and responses are typed data
 * classes that mirror the JSON documents of the API
 * (`io.github.pt9912.pgchangefeed.http.model`); every non-success response
 * becomes a typed [PgChangeFeedException] subtype instead of a raw HTTP status
 * code mixed with the success path. Connection errors and timeouts of the
 * transport are not converted; they reach the caller as the `IOException` of
 * `java.net.http.HttpClient`.
 *
 * The `java.net.http.HttpClient` passed to the public constructor is passed
 * in, not owned — the caller controls its lifetime, connection pooling and any
 * redirect/proxy configuration; this class never closes it. The bearer token
 * and server address come from [PgChangeFeedClientOptions], supplied at
 * construction — there is no global or static state, so a process can hold
 * several independently configured instances at once.
 */
class PgChangeFeedHttpClient internal constructor(
    private val transport: HttpTransport,
    private val options: PgChangeFeedClientOptions,
) {
    /**
     * Public constructor — wraps the caller's `java.net.http.HttpClient` in
     * the real [JdkHttpTransport]; see [HttpTransport] for why the client
     * depends on a transport interface rather than on `HttpClient` directly.
     */
    constructor(httpClient: HttpClient, options: PgChangeFeedClientOptions) :
        this(JdkHttpTransport(httpClient), options)

    /**
     * Registers a consumer, a named reader whose position the server keeps
     * (`POST /consumers`, admin token). Registering an existing consumer
     * changes nothing; the response reports it with `alreadyRegistered`.
     */
    fun registerConsumer(request: RegisterConsumerRequest): RegisterConsumerResponse =
        post("/consumers", gson.toJson(request))

    /**
     * Stores the position up to which a consumer has processed a source
     * (`POST /consumers/acknowledge`, admin token). Repeating the stored
     * position has no effect; an earlier position, or a position of another
     * source, is rejected with [PgChangeFeedBadRequestException].
     */
    fun acknowledgeConsumer(request: AcknowledgeConsumerRequest): AcknowledgeConsumerResponse =
        post("/consumers/acknowledge", gson.toJson(request))

    /**
     * Reads the stored position of a consumer (`GET /consumers/position`,
     * reader or admin token). `acknowledged` is `false` for a consumer that
     * never acknowledged; `offset` is then the starting position.
     */
    fun getConsumerPosition(consumerId: String): ConsumerPositionResponse =
        get("/consumers/position?" + buildQuery("consumer_id" to consumerId))

    /**
     * Removes a consumer (`POST /consumers/remove`, admin token); `removed` is
     * `false` for one that was never registered.
     */
    fun removeConsumer(consumerId: String): RemoveConsumerResponse =
        post("/consumers/remove", gson.toJson(mapOf("consumer_id" to consumerId)))

    /**
     * Starts capturing a table of a source (`POST /tables/enable`, admin
     * token). `alreadyEnabled` is `true` when the table was captured already; a
     * table that does not exist in the source database raises
     * [PgChangeFeedNotFoundException].
     */
    fun enableTable(request: EnableTableRequest): EnableTableResponse =
        post("/tables/enable", gson.toJson(request))

    /**
     * Stops capturing a table (`POST /tables/disable`, admin token). `retained`
     * is `true` when changes already stored for the table remain readable; a
     * table that does not exist in the source database raises
     * [PgChangeFeedNotFoundException].
     */
    fun disableTable(request: DisableTableRequest): DisableTableResponse =
        post("/tables/disable", gson.toJson(request))

    /**
     * Tells whether a table is captured (`enabled`) or no longer captured with
     * stored changes remaining (`retained`) (`GET /tables/status`, reader or
     * admin token). A table that was never enabled reports both as `false`; a
     * table that does not exist in the source database raises
     * [PgChangeFeedNotFoundException].
     */
    fun getStatus(source: String, schema: String, table: String, publication: String): TableStatusResponse =
        get(
            "/tables/status?" + buildQuery(
                "source" to source,
                "schema" to schema,
                "table" to table,
                "publication" to publication,
            ),
        )

    /**
     * Lists the captured tables and the tables that are no longer captured but
     * whose stored changes remain (`GET /tables`, reader or admin token).
     */
    fun listTables(source: String, publication: String): ListTablesResponse =
        get("/tables?" + buildQuery("source" to source, "publication" to publication))

    /**
     * Deletes the stored changes of a source that are older than `minAgeNanos`
     * and that every consumer with a stored position has passed
     * (`POST /retention/run`, admin token); `deleted` is the number removed.
     */
    fun runRetention(request: RunRetentionRequest): RunRetentionResponse =
        post("/retention/run", gson.toJson(request))

    /**
     * Reads stored changes of a source (`GET /changes`, reader or admin
     * token). [source] is mandatory; [schema]/[table] are optional and
     * independent; [from] is inclusive, [to] exclusive (both `commit_position`
     * values). [limit] cuts rows, not positions, and there is no default
     * limit. A range without changes returns an empty list.
     */
    fun readChanges(
        source: String,
        schema: String? = null,
        table: String? = null,
        from: Long? = null,
        to: Long? = null,
        limit: Int? = null,
    ): ReadChangesResponse {
        val query = buildQuery(
            "source" to source,
            "schema" to schema,
            "table" to table,
            "from" to from?.toString(),
            "to" to to?.toString(),
            "limit" to limit?.toString(),
        )
        return get("/changes?$query")
    }

    private fun buildQuery(vararg params: Pair<String, String?>): String =
        params
            .filter { (_, value) -> value != null }
            .joinToString("&") { (key, value) -> "$key=${encode(value!!)}" }

    private inline fun <reified T> get(pathAndQuery: String): T =
        dispatch("GET", pathAndQuery, body = null)

    private inline fun <reified T> post(path: String, body: String): T =
        dispatch("POST", path, body)

    private inline fun <reified T> dispatch(method: String, pathAndQuery: String, body: String?): T {
        val headers = buildMap {
            put("Authorization", "Bearer ${options.apiToken}")
            if (body != null) {
                put("Content-Type", "application/json; charset=utf-8")
            }
        }
        val request = TransportRequest(
            method = method,
            url = options.address.toString().trimEnd('/') + pathAndQuery,
            headers = headers,
            body = body,
        )
        val response = transport.send(request)
        if (response.statusCode !in 200..299) {
            throw buildException(response.statusCode, response.body)
        }
        return parseSuccessBody(response.statusCode, response.body)
    }

    private inline fun <reified T> parseSuccessBody(statusCode: Int, body: String): T {
        val result = try {
            gson.fromJson(body, T::class.java)
        } catch (ex: JsonSyntaxException) {
            throw PgChangeFeedMalformedResponseException(
                statusCode,
                "PG Change Feed HTTP API returned status $statusCode with a response body that is " +
                    "not valid JSON.",
                ex,
            )
        }
        return result
            ?: throw PgChangeFeedMalformedResponseException(
                statusCode,
                "PG Change Feed HTTP API returned status $statusCode with an empty or null response " +
                    "body.",
            )
    }

    private fun buildException(statusCode: Int, body: String): PgChangeFeedException {
        val message = extractErrorMessage(body)
        return when (statusCode) {
            400 -> PgChangeFeedBadRequestException(statusCode, message)
            401 -> PgChangeFeedUnauthorizedException(statusCode, message)
            403 -> PgChangeFeedForbiddenException(statusCode, message)
            404 -> PgChangeFeedNotFoundException(statusCode, message)
            500 -> PgChangeFeedServerErrorException(statusCode, message)
            else -> PgChangeFeedUnexpectedStatusException(statusCode, message)
        }
    }

    private fun extractErrorMessage(body: String): String =
        try {
            gson.fromJson(body, ErrorResponse::class.java)?.error ?: body
        } catch (ex: JsonSyntaxException) {
            body
        }

    /**
     * RFC 3986 percent-encoding (unreserved: `A-Za-z0-9-_.~`) — not
     * `java.net.URLEncoder`'s form-encoding (which encodes a space as `+`
     * instead of `%20`).
     */
    private fun encode(value: String): String {
        val builder = StringBuilder()
        for (byte in value.toByteArray(StandardCharsets.UTF_8)) {
            val unsigned = byte.toInt() and 0xFF
            val c = unsigned.toChar()
            val isUnreserved = c in 'A'..'Z' || c in 'a'..'z' || c in '0'..'9' ||
                c == '-' || c == '_' || c == '.' || c == '~'
            if (isUnreserved) {
                builder.append(c)
            } else {
                builder.append('%')
                builder.append(String.format("%02X", unsigned))
            }
        }
        return builder.toString()
    }

    private companion object {
        val gson = Gson()
    }
}
